src/Microsoft.ML.TensorFlow/TensorflowTransform.cs

```diff
@@ -111,34 +102,55 @@ private static TensorFlowTransform Create(IHostEnvironment env, ModelLoadContext
+
             // *** Binary format ***
+            // byte: indicator for frozen models
             // stream: tensorFlow model.
             // int: number of input columns
             // for each input column
             //   int: id of int column name
             // int: number of output columns
             // for each output column
             //   int: id of output column name
-            byte[] modelBytes = null;
-            if (!ctx.TryLoadBinaryStream("TFModel", r => modelBytes = r.ReadByteArray()))
-                throw env.ExceptDecode();
-            var numInputs = ctx.Reader.ReadInt32();
-            env.CheckDecode(numInputs > 0);
-            string[] inputs = new string[numInputs];
-            for (int j = 0; j < inputs.Length; j++)
-                inputs[j] = ctx.LoadNonEmptyString();
+            GetModelInfo(env, ctx, out string[] inputs, out string[] outputs, out bool isFrozen);
+            if (isFrozen)
+            {
+                byte[] modelBytes = null;
+                if (!ctx.TryLoadBinaryStream("TFModel", r => modelBytes = r.ReadByteArray()))
+                    throw env.ExceptDecode();
 
-            bool isMultiOutput = ctx.Header.ModelVerReadable >= 0x00010002;
-            var numOutputs = 1;
-            if (isMultiOutput)
-                numOutputs = ctx.Reader.ReadInt32();
+                return new TensorFlowTransform(env, LoadTFSession(env, modelBytes), inputs, outputs, isFrozen);
+            }
 
-            env.CheckDecode(numOutputs > 0);
-            var outputs = new string[numOutputs];
-            for (int j = 0; j < outputs.Length; j++)
-                outputs[j] = ctx.LoadNonEmptyString();
+            var tempDirPath = Path.GetFullPath(Path.Combine(Path.GetTempPath(), RegistrationName + "_" + Guid.NewGuid()));
+            TensorFlowUtils.CreateTempDirectory(tempDirPath);
+
+            var load = ctx.TryLoadBinaryStream("TFSavedModel", br =>
+            {
+                int count = br.ReadInt32();
+                for (int n = 0; n < count; n++)
+                {
+                    string relativeFile = br.ReadString();
+                    long fileLength = br.ReadInt64();
+
+                    string fullFilePath = Path.Combine(tempDirPath, relativeFile);
+                    string fullFileDir = Path.GetDirectoryName(fullFilePath);
+
+                    if (fullFileDir != tempDirPath)
+                    {
+                        TensorFlowUtils.CreateTempDirectory(fullFileDir);
+                    }
 
-            return new TensorFlowTransform(env, modelBytes, inputs, outputs);
+                    using (var fs = new FileStream(fullFilePath, FileMode.Create, FileAccess.Write))
+                    {
+                        long actualRead = br.BaseStream.CopyRange(fs, fileLength);
+                        env.Assert(actualRead == fileLength);
+                    }
+                }
+            });
+
+            var session = GetSession(env, tempDirPath);
```

Comment:
>GetSession [](start = 26, length = 10)

This call can be used for the isFrozen case too. Might look simpler like this:

if (!isFrozen)
{
// do everything in lines 125-150
}

var session = GetSession(...);
return new TensorFlowTransform(session,...);

Reply:
This way you can have only one ctor, that takes a session and all the other information it needs - you can pass the directory name or null to indicate frozen, and a bool for isTempDir.

---
In reply to: [219215618](https://github.com/dotnet/machinelearning/pull/853#discussion_r219215618) [](ancestors = 219215618)

Reply:
I refactored the code a bit, so we now have 1 constructor.

btw..for the frozen case we are not calling GetSession, we are calling  LoadTFSession(env, modelBytes) to load the session from bytes. 

---
In reply to: [219216429](https://github.com/dotnet/machinelearning/pull/853#discussion_r219216429) [](ancestors = 219216429,219215618)